### PR TITLE
Streamline key management for build caches

### PR DESCRIPTION
**Pull request by opadron**
Fixes the logic Spack uses when getting keys from a mirror, which was broken after merging #15002 and #15361.

- ~Update `spack.util.web.list_url()` to iterate recursively over all files~
   - ~Changed to be more consistent across `s3://` and `file://` URLs.~

 - ~Changed `bindist.get_keys()` so that it no longer depends on an index.~
   - ~Instead, it will find *all* files under the prefix URL that end
     with .key or .pub~

 - Simplified `bindist.get_keys()` so that it uses the web/url abstraction
   instead of separate code paths for `s3://` and `file://` resources.

TODO:

- [x] Fix tests
- [x] Confirm getting keys from mirror is working
- [x] Add new tests covering changes

EDIT: Crossed out parts of this PR that has since changed.  See discussion for details.

### CODE REVIEW

**Comment by opadron** [comment]
Tagging with "don't-merge-yet": There is still more work to be done here with respect to basic key management. (per conversations with @tgamblin)

**Comment by scottwittenburg** [comment]
@opadron Quick question: It looks like this is only dealing with the read side of things, do we already have nice commands for adding, say, a public key to a buildcache?

Attn: @eugeneswalker

**Comment by gartung** [comment]
I never added that ability add a public key to the buildcache tarfile.

I also have an issue opened to add signing of buildcaches after they are created.
https://github.com/spack/spack/issues/13732

**Comment by gartung** [comment]
Or better yet, move the downloading and trusting of keys to the spack gpg command.

**Comment by scottwittenburg** [comment]
> I never added that ability add a public key to the buildcache tarfile.

Sorry I wasn't clear there.  I meant to ask about just uploading a public key to the buildcache directory, for example, into the directory where this PR will look for them.  I think in a recent discussion were thinking of expecting keys to live under the buildcache directory in a folder named `_keys`.  That folder would also contain a simple json index of the available keys.  And the command we might introduce for adding such a key would also handle updating that index.

**Comment by eugeneswalker** [comment]
I'm not sure I understand exactly the plan here, but if it is to list all objects in a directory that could potentially contain many thousands of files in an attempt to find those that have .key or .pub extensions, I am against that. Are we talking about a specific directory like `_keys` where only keys would be stored?

**Comment by scottwittenburg** [comment]
> the command we might introduce for adding such a key would also handle updating that index.

It might also be nice, when we're adding a buildcache entry to a mirror via `spack buildcache create`, to just make sure the public part of the key is uploaded and indexed as well.

**Comment by scottwittenburg** [comment]
> Are we talking about a specific directory like `_keys` where only keys would be stored?

Yes, IIRC that's currently the plan, although I don't see that notion reflected here yet.

**Comment by opadron** [comment]
That's right.  The plan for this PR is to pivot on the original idea and move to a `_keys` directory under the build cache that will have all the keys and a json index.  This PR will also include plumbing for uploading the keys to the build cache (and subsequently updating the index).

**Comment by opadron** [comment]
For `spack.util.web.list_url()`, the changes are moot for this PR since we're moving away from programmatic listings and relying on an index.  I will probably end up reverting that part of the change set.

**Comment by opadron** [comment]
PTAL

I've started reworking how keys are managed for build caches.  Please see the commit messages for these last two commits for details.  Please weigh in on whether you like the direction the changes are going.

**Review by gartung — APPROVED** [review]

**Comment by opadron** [comment]
PTAL.  I've reproduced a summary of the major changes below:

> Switches to maintaining a build cache's keys under build_cache/_pgp.
Within this directory is an index.json file listing all the available
keys and a <fingerprint>.pub file for each such key.

> New spack subcommand: spack gpg publish

EDIT: ~This is ready to merge.~  Actually, I'll rebase real quick and then it'll be ready!

**Comment by eugeneswalker** [comment]
Using this PR I was able to:
* add a key to a local filesystem build cache: `spack gpg publish -d <path> <key-id>`
* generate an key index: `spack gpg publish --rebuild-index -d <path>`
* fetch and trust the key: `spack buildcache keys -it`

**Comment by opadron** [comment]
Update: I'm reworking some of the internal plumbing regarding how `gpg` is called by Spack.  The hope is that these changes will make a fix for the mac os test failure more straightforward.

**Comment by opadron** [comment]
@gartung (or anyone else who wants to review):

Could you take another pass at this?  I've had to add more changes since your last review.

**Comment by gartung** [comment]
@eugeneswalker Can you take a look. I don't have any mirrrors up at the moment to test this.

File: lib/spack/spack/binary_distribution.py [hunk 1]
```diff
@@ -293,7 +292,8 @@ def generate_package_index(cache_prefix):
     file_list = (
         entry
         for entry in web_util.list_url(cache_prefix)
-        if entry.endswith('.yaml'))
+        if entry.endswith('.yaml')
+        and os.sep not in entry)
```
Comment (scottwittenburg) [inline lib/spack/spack/binary_distribution.py]: What's the extra condition about?  Is it because `list_url` is now going to iterate the mirror contents recursively?

File: lib/spack/spack/binary_distribution.py [hunk 2]
```diff
@@ -898,41 +948,46 @@ def get_specs():
...
         tty.die("Please add a spack mirror to allow " +
                 "download of build caches.")
 
-    keys = set()
+    for mirror in mirror_collection.values():
+        fetch_url = mirror.fetch_url
+        keys_url = url_util.join(fetch_url,
+                                 _build_cache_relative_path,
+                                 _build_cache_keys_relative_path)
+        keys_index = url_util.join(keys_url, 'index.json')
```
Comment (scottwittenburg) [inline lib/spack/spack/binary_distribution.py:964]: Your PR description says one of the three things this PR does is update `bindist.get_keys()` so it no longer uses an index, and instead finds all files matching certain criteria under the `cache_prefix`.  But that doesn't look like what's going on here.

File: lib/spack/spack/binary_distribution.py [hunk 3]
```diff
@@ -468,7 +509,15 @@ def build_tarball(spec, outdir, force=False, rel=False, unsigned=False,
               .format(spec, remote_spackfile_path))
 
     try:
-        # create an index.html for the build_cache directory so specs can be
+        # push the key to the build cache's _pgp directory so it can be
+        # imported
+        if not unsigned:
+            push_keys(outdir,
```
Comment (scottwittenburg) [inline lib/spack/spack/binary_distribution.py:514]: This is nice, if I understand correctly.  When we create a tarball and push it to a remote mirror, we will also push the key it was signed with?  I haven't got to that bit of your change yet, I'm just assuming it somehow only pushes the public part of the key.

File: lib/spack/spack/test/cmd/pkg.py [hunk 4]
```diff
@@ -122,7 +122,8 @@ def test_pkg_add(mock_pkg_git_repo):
     with working_dir(mock_pkg_git_repo):
         try:
             assert ('A  pkg-e/package.py' in
-                    git('status', '--short', output=str))
+                    git('-c', 'color.status=off',
```
Comment (scottwittenburg) [inline lib/spack/spack/test/cmd/pkg.py]: Just curious: this seems unrelated, what's going on here?

File: lib/spack/spack/test/stage.py [hunk 5]
```diff
@@ -833,7 +833,8 @@ def test_get_stage_root_in_spack(self, clear_stage_root):
                 assert 'spack' in path.split(os.path.sep)
 
                 # Make sure cached stage path value was changed appropriately
-                assert spack.stage._stage_root == test_path
+                assert spack.stage._stage_root in (
+                    test_path, os.path.join(test_path, getpass.getuser()))
```
Comment (scottwittenburg) [inline lib/spack/spack/test/stage.py:837]: Again, just curious, seems unrelated?

File: lib/spack/spack/util/gpg.py [hunk 6]
```diff
@@ -3,20 +3,73 @@
...
+    Otherwise, if the environment variable "SPACK_GNUPGHOME" is set, then that
+    value will be used.
+
+    Otherwise, the default gpg path for Spack will be used.
+
+    See also: gnupg_home_override()
+    """
+    return (gnupg_home or
+            _gnupg_home_override or
+            os.getenv('SPACK_GNUPGHOME') or
```
Comment (scottwittenburg) [inline lib/spack/spack/util/gpg.py:44]: Do you need a default of `None` here?  Like `os.getenv('SPACK_GNUPGHOME', None)`...

File: lib/spack/spack/util/gpg.py [hunk 7]
```diff
@@ -31,43 +84,255 @@ def parse_keys_output(output):
...
+        self.gnupg_home = get_gnupg_home(gnupg_home)
+
+    @cached_property
+    def prep(self):
+        # Make sure that suitable versions of gpgconf and gpg are available
+        ensure_gpg()
+
+        # Make sure that the GNUPGHOME exists
+        if not os.path.exists(self.gnupg_home):
+            os.makedirs(self.gnupg_home)
```
Comment (scottwittenburg) [inline lib/spack/spack/util/gpg.py:281]: Did you omit the `chmod 700 ...` on purpose here?

File: lib/spack/spack/util/gpg.py [hunk 8]
```diff
@@ -83,64 +348,105 @@ def create(cls, **kwargs):
...
+                return _callable(*args, **kwargs)
+
+        result.name = name
+        return result
+
+    for func, name in wrap:
+        namespace[name] = _make_wrapped_callable(func, name)
+
+
+_make_wrapped_callables(globals())
```
Comment (scottwittenburg) [inline lib/spack/spack/util/gpg.py]: My $0.02: I'm not a fan of the meta-programming approach taken here, seems like a lot of complexity for the benefit of users not having to create an instance of the `Gpg` class.  It took a non-zero amount of time and effort to figure out what you were doing with the various bits: the mysterious and uncommented `wrap` decorator class that inherits from `list`, its global instance used in decorating selected methods of the `Gpg` class, the definition and subsequent deletion of this method where you generate the module level handles from stuff stored by the decorator, etc...

Again, just my opinion, but I found all that stuff kind of a pain to review, and I would have preferred something straightforward like explicit definition of the handful of module-level methods you want to expose, in terms of a singleton `Gpg` instance.  Maybe I'm just exposing myself as simple-minded here though.

**Review by scottwittenburg — APPROVED** [review]
I haven't actually tested this code myself, but just looked through it and provided some comments and a few questions.  Seems to make sense at a high level, and it's fixing some broken things we really need fixed, so I'm going to approve in advance of any back and forth that may happen as a result of my review comments.  Thanks @opadron!

File: lib/spack/spack/binary_distribution.py (hunk 1 above)
Comment (opadron) [reply lib/spack/spack/binary_distribution.py]: I think this extra check is not necessary, I will remove it and see if the tests still pass.

File: lib/spack/spack/binary_distribution.py (hunk 3 above)
Comment (opadron) [reply lib/spack/spack/binary_distribution.py:514]: This is correct.  See `build_distribution.push_keys()` for the details.  In particular, take note of lines:

`1019` : `keys = spack.util.gpg.public_keys(*(keys or []))`

and

`1050` : `spack.util.gpg.export_keys(export_target, fingerprint)`

... the latter of which calls `gpg --export`, which is never supposed to export secret subkeys (for that, there is an entirely separate `gpg` command).

File: lib/spack/spack/binary_distribution.py (hunk 2 above)
Comment (opadron) [reply lib/spack/spack/binary_distribution.py:964]: That's my mistake.  The description documented the original intent of the PR, which has since changed.

File: lib/spack/spack/test/cmd/pkg.py (hunk 4 above)
Comment (opadron) [reply lib/spack/spack/test/cmd/pkg.py]: Here, I was trying to work around an issue in the tests where my local git config was causing failures.

After taking a closer look and doing some googling, I realize that running git as part of a script (or in this case, as part of Spack) and *not* taking the extra measures needed to ensure that it is isolated from user- and system-provided git configs is inherently broken.  This is a separate issue that I intend to follow up on with a more-comprehensive PR.

For this PR, I will revert the change.

File: lib/spack/spack/test/stage.py (hunk 5 above)
Comment (opadron) [reply lib/spack/spack/test/stage.py:837]: This works around an issue on OSX.  I don't remember the details, so I can't be sure that there isn't a deeper issue that this change masks.  Maybe this warrants more attention?

Comment (scottwittenburg) [reply lib/spack/spack/test/stage.py:837]: Up to you.  I didn't really intend to hold you up on this PR, but rather just try to understand what you were doing.

File: lib/spack/spack/util/gpg.py (hunk 6 above)
Comment (opadron) [reply lib/spack/spack/util/gpg.py:44]: I don't think it is necessary.

> os.getenv(varname[, value])
Return the value of the environment variable varname if it exists, or value if it doesn’t. value defaults to None.

There's no note of any version of python where they made value optional, which leads me to believe that it's always been that way.

File: lib/spack/spack/util/gpg.py (hunk 7 above)
Comment (opadron) [reply lib/spack/spack/util/gpg.py:281]: Thanks for that catch.  Fixed in upcoming commits.

File: lib/spack/spack/util/gpg.py (hunk 8 above)
Comment (opadron) [reply lib/spack/spack/util/gpg.py]: Not at all!  I hate it, too (and I'm the one who wrote it!).

I think there's a neat meta programming idea here that could be pulled out and stashed away in some module somewhere, but frankly I was just going for future-proofing, and sometimes I just need someone to point out when I'm being too clever for my own good!

I agree this should be simpler.  I'll take another pass and see how it turns out.

Comment (opadron) [reply lib/spack/spack/util/gpg.py]: Have a look at the latest commit.  Is that easier to make sense of?
